packages/core/snapshot/manager.go

```diff
@@ -0,0 +1,175 @@
+package snapshot
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/iotaledger/goshimmer/packages/core/epoch"
+	"github.com/iotaledger/goshimmer/packages/core/ledger"
+	"github.com/iotaledger/goshimmer/packages/core/notarization"
+	"github.com/iotaledger/goshimmer/packages/core/tangleold"
+	"github.com/iotaledger/goshimmer/packages/node/database"
+	"github.com/iotaledger/hive.go/core/generics/event"
+	"github.com/iotaledger/hive.go/core/kvstore"
+)
+
+const (
+	prefixSolidEntryPoint byte = iota
+)
+
+// Manager is the snapshot manager.
+type Manager struct {
+	tangle          *tangleold.Tangle
+	notarizationMgr *notarization.Manager
+
+	baseStore kvstore.KVStore
+	sepsLock  sync.RWMutex
+}
+
+// NewManager creates and returns a new snapshot manager.
+func NewManager(store kvstore.KVStore, t *tangleold.Tangle, nmgr *notarization.Manager) (new *Manager) {
+	new = &Manager{
+		tangle:          t,
+		notarizationMgr: nmgr,
+	}
+
+	new.baseStore = store
+
+	new.tangle.ConfirmationOracle.Events().BlockAccepted.Attach(event.NewClosure(func(e *tangleold.BlockAcceptedEvent) {
+		e.Block.ForEachParent(func(parent tangleold.Parent) {
+			index := parent.ID.EpochIndex
+			if index < e.Block.EI() {
+				new.insertSolidEntryPoint(parent.ID)
+			}
+		})
+	}))
+
+	new.tangle.ConfirmationOracle.Events().BlockOrphaned.Attach(event.NewClosure(func(event *tangleold.BlockAcceptedEvent) {
+		new.removeSolidEntryPoint(event.Block, event.Block.LatestConfirmedEpoch())
+	}))
+
+	return
+}
+
+// CreateSnapshot creates a snapshot file from node with a given name.
+func (m *Manager) CreateSnapshot(snapshotFileName string) (header *ledger.SnapshotHeader, err error) {
+	ecRecord, lastConfirmedEpoch, err := m.tangle.Options.CommitmentFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	// lock the entire ledger in notarization manager until the snapshot is created.
+	m.notarizationMgr.WriteLockLedger()
+	defer m.notarizationMgr.WriteUnlockLedger()
+
+	// lock the entire solid entry points storage until the snapshot is created.
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	headerPord := func() (header *ledger.SnapshotHeader, err error) {
+		header = &ledger.SnapshotHeader{
+			FullEpochIndex: lastConfirmedEpoch,
```

Comment:
Currently our snapshottable ledgerstate it is not fixed at our latestConfirmedEpoch, take a look at the `newEpochRoots`:
```
	// We advance the LedgerState to the next epoch.
	f.commitLedgerState(ei - epoch.Index(f.snapshotDepth))
```

This bit is invoked whenever we are able to advance our latest committable epoch.

Reply:
Until we have a confirmation gadget you have to make sure that FullEpochIndex here is the same one we have the Notarization manager's LedgerState on!